Vector drafts

DrewKimball opened the pull request.

DrewKimball committed:
vecdesc: add vector index config to the index descriptor

This patch introduces the vecdesc package (to avoid circular
dependencies) and populates it with the vector index config descriptor.

Epic: CRDB-42943

DrewKimball committed:
sql/vecindex: add vector index manager

This commit adds a `vecindex.Manager` to each SQL process, which manages
the `vecindex.VectorIndex` for each active vector index in a given node.
Users within SQL execution can access a `VectorIndex` via `Manager.Get()`.

Epic: CRDB-42943

DrewKimball committed:
add vector index support to schema changers

DrewKimball committed:
support vector index mutations

mw5h committed:
vecstore: change to using IDs instead of descriptors

Previously, the persistent data store took a table descriptor and
catalog.Index to describe the table and index of the associated vector
index. This patch changes those to descriptor IDs and then fetches the
needed information. This resolves an impedence mismatch with the vector
manager, which deals in descriptor IDs and avoids using stale
descriptors.

This patch also provides a nil implementation of MergeStats, which is
called by the vector index, but for which a nil response is okay (for
now).

Epic: CRDB-42943

DrewKimball committed:
opt: rename VectorPartitionSearch

This commit renames the `VectorPartitionSearch` operator to
`VectorMutationSearch`. This reflects the fact that the operator is
used to prepare the information needed for index updates.

Epic: None

Release note: None

DrewKimball committed:
opt: modify VectorMutationSearch to produce quantized vectors

This commit changes `VectorMutationSearch` so that it produces quantized
and encoded vectors, rather than partition centroids. This shifts the
responsibility of quantizing vectors to the `VectorMutationSearch`
operator rather than the insert/update code paths.

Epic: CRDB-42943

Release note: None

mw5h committed:
vecindex: create a unit test for the vector index manager

This patch introduces a unit test for the vector index manager and
testing knobs for the vector index.

Epic: CRDB-42943

DrewKimball committed:
sql: plumb vector search into execution logic

DrewKimball committed:
sql: initialize the root partition upon creation of a vector index

DrewKimball committed:
fixup sql: plumb vector search into execution logic

DrewKimball committed:
fixup support vector index mutations

DrewKimball committed:
vecindex: some bug fixes and todos

DrewKimball committed:
sql: add builtin functions for testing vector indexes

This commit adds two helper builtin functions for internal testing of
vector indexes:
`crdb_internal.display_vector_index`: prints a tree-formatted string
representing the index structure.
`crdb_internal.process_vector_index_fixups`: waits until the fixup queue
for a vector index is empty.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/139992)
<!-- Reviewable:end -->

DrewKimball committed:
opt: project the vector column in GenerateVectorSearch

This commit fixes a bug in `GenerateVectorSearch` that caused an
internal error in some indexed vector-search queries during planning.
The issue was that the index-join added to the plan didn't include
the vector column, which is needed to project the distance column
after the vector search operator.

DrewKimball committed:
sql: make EXPLAIN ANALYZE work for vector search operators

DrewKimball committed:
sql: queue fixups for vector index puts

DrewKimball committed:
vecindex: correctly roll back txn after a failed commit

DrewKimball commented:
This is superseded by #140770, but I'll leave it for now as a stable working prototype.
